### DISCUSSION TITLE
WASM in Extension: WASI vs WASI-less

### DISCUSSION THREAD

**Discussion by sourishkrout** [microsoft/vscode-discussions#211] [Extension Development QnA] [answered]
Hello everybody! Finally had some time to sift through https://github.com/microsoft/vscode-wasm plus its testbed examples. Great work! Really nifty how one is able to delegate WASM execution to worker threads. Great work @dbaeumer and team!

The question I couldn't quite resolve for myself is this: Can I instantiate and run WASM directly inside of the extension (host) if it won't require any WASI/sys interfaces? E.g. cross-compiling processing biz logic to avoid duplication? I mean it works, I've tested it. Would like to know if it will continue to work.

I can totally see how `vscode-wasm` is geared towards making a stripped down Python interpreter or the likes available in a browser (no extension host). I suppose one could always run WASM in a worker thread as long as `ServiceConnection` allows custom messages to be passed back and forth (is that possible?). For pure computational logic it add overhead which is acceptable for some but maybe not for other use-cases.

Any insights how `vscode-wasm` should or should not be used in contrast to WASI-less use-cases (computational model) as well as general recommendations would be much appreciate it. Thank you!

**Comment by dbaeumer** [comment, marked as answer]
You might want to have a look at https://github.com/microsoft/vscode-python-web-wasm which does uses the vscode-wasm libs with custom messages. Although the repository has web in its name the extension will perfectly run in a Desktop extension host as well. When executed in the browser it is run in an extension host as well with the difference that the extension host is executed in a web worker instead a node process.

**Reply by sourishkrout** [reply]
Thank you @dbaeumer! After a little struggle with some file output locations (documented here: https://github.com/microsoft/vscode-python-web-wasm/issues/17), I've been able to get https://github.com/microsoft/vscode-python-web-wasm up and running.

I will take some time to work through it and reported back with conclusions or questions.

**Comment by zzz6519003** [comment]
@dbaeumer LSP is awesome, thx from China!
